File: ios/SpentyAI/SpentyAI/Features/AIChat/AIChatViewModel.swift
import Foundation
import SwiftUI

@Observable
final class AIChatViewModel {

    // MARK: - State

    var messages: [ChatMessage] = []
    var input: String = ""
    var isSending = false
    var suggestions: [String] = []
    var errorMessage: String?
    var scrollToBottomTrigger = 0
    var scrollToLastMessageTopTrigger = 0

    // MARK: - Voice State

    /// Whether voice responses (text-to-speech) are enabled.
    var isVoiceResponseEnabled: Bool = false

    /// Whether fullscreen voice mode is active (continuous listen + auto-speak).
    var isVoiceModeActive: Bool = false

    /// The speech manager instance.
    let speechManager = SpeechManager()

    // MARK: - Dependencies

    private let repository: AIChatRepository

    // MARK: - Init

    init(repository: AIChatRepository = .shared) {
        self.repository = repository
    }

    // MARK: - Computed

    var canSend: Bool {
        // The button is sendable when there's text to send AND we're not
        // already mid-request. "Text to send" comes from either the visible
        // input field OR the live speech transcript — the latter covers the
        // race where the user taps Send before the .onChange live-mirror has
        // propagated the latest words into the input. sendMessage() then
        // rescues the transcript if input is still empty when it actually
        // reads it.
        let inputEmpty = input.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        let transcribedEmpty = speechManager.transcribedText
            .trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        return !isSending && !(inputEmpty && transcribedEmpty)
    }

    // MARK: - Actions

    @MainActor
    func sendMessage() async {
        // Reentry guard: a fast double-tap on the Send button or a voice-mode
        // tap arriving while a previous request is still in flight would queue
        // two requests and double-fire. Protect at the source — `isSending`
        // is read by the UI to disable the button but UI redraw has a frame
        // of latency; guard here too.
        guard !isSending else {
            #if DEBUG
            print("⚠️ AIChat sendMessage ignored — already sending")
            #endif
            return
        }

        // Belt-and-suspenders: if `input` is empty but the user is mid-dictation
        // (mic on with a transcript), fall through to the transcribed text and
        // also stop the mic. This catches a race where the user taps Send before
        // .onChange has propagated the latest transcribedText into input.
        if input.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            let fallback = speechManager.transcribedText.trimmingCharacters(in: .whitespacesAndNewlines)
            if !fallback.isEmpty {
                #if DEBUG
                print("ℹ️ AIChat sendMessage rescuing from transcribedText (input was empty)")
                #endif
                if speechManager.isListening { speechManager.stopListening() }
                input = fallback
                speechManager.resetTranscription()
            }
        }

        let text = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else {
            #if DEBUG
            print("⚠️ AIChat sendMessage ignored — input is empty after fallback")
            #endif
            return
        }

        #if DEBUG
        print("📤 AIChat sendMessage start — \(text.prefix(60))")
        #endif

        // Add user message immediately
        let userMessage = ChatMessage(
            id: UUID().uuidString,
            role: "user",
            content: text,
            transactionPosted: nil,
            transaction: nil,
            invoiceCreated: nil,
            invoice: nil,
            billCreated: nil,
            bill: nil
        )
        messages.append(userMessage)
        input = ""
        isSending = true
        errorMessage = nil
        scrollToBottom()

        do {
            let response = try await repository.sendMessage(text, conversation: messages)
            messages.append(response)
            scrollToLastMessageTop()

            // If the AI just created an account, fire a refresh notification
            // so any open AccountListView (and dashboard balance widgets) pick
            // it up without the user having to pull-to-refresh. Mirrors the
            // pattern used for AI-posted transactions.
            if response.accountCreated == true {
                NotificationCenter.default.post(name: .accountsDidChange, object: nil)
            }

            // Speak the response if voice responses are enabled or voice mode is active
            if (isVoiceResponseEnabled || isVoiceModeActive), let content = response.content, !content.isEmpty {
                speechManager.speak(content)
            }

            // In voice mode, restart listening after the AI finishes responding
            if isVoiceModeActive {
                // Wait briefly for TTS to finish before re-listening
                Task { @MainActor in
                    // Give TTS a moment to start, then wait for it to finish
                    try? await Task.sleep(for: .milliseconds(500))
                    while speechManager.isSpeaking {
                        try? await Task.sleep(for: .milliseconds(200))
                    }
                    // Guard: user may have exited voice mode while we were
                    // waiting for TTS — don't restart listening if so.
                    guard isVoiceModeActive else {
                        #if DEBUG
                        print("ℹ️ AIChat voice-mode auto-restart aborted (user exited)")
                        #endif
                        return
                    }
                    speechManager.resetTranscription()
                    speechManager.startListening()
                }
            }
            #if DEBUG
            print("✅ AIChat sendMessage success")
            #endif
        } catch let error as APIError {
            #if DEBUG
            print("❌ AIChat sendMessage APIError: \(error.localizedDescription)")
            #endif
            errorMessage = error.localizedDescription
        } catch {
            #if DEBUG
            print("❌ AIChat sendMessage error: \(error.localizedDescription)")
            #endif
            errorMessage = "Failed to get a response. Please try again."
        }

        isSending = false
    }

    @MainActor
    func sendSuggestion(_ suggestion: String) async {
        input = suggestion
        await sendMessage()
    }

    @MainActor
    func loadHistory() async {
        do {
            messages = try await repository.loadHistory()
            scrollToBottom()
        } catch {
            // Loading history is OPPORTUNISTIC. If it fails (network blip,
            // first-time user with no history yet, transient decoding hiccup,
            // anything) the user can still send messages — surfacing a
            // blocking alert here was over-aggressive and hid the chat UI.
            // Log for debugging; leave messages empty.
            #if DEBUG
            print("⚠️ AIChat loadHistory failed: \(error.localizedDescription)")
            #endif
        }
    }

    @MainActor
    func clearHistory() async {
        do {
            try await repository.clearHistory()
            messages.removeAll()
            errorMessage = nil
        } catch let error as APIError {
            errorMessage = error.localizedDescription
        } catch {
            errorMessage = "Could not clear history. Please try again."
        }
    }

    @MainActor
    func loadSuggestions() async {
        do {
            suggestions = try await repository.loadSuggestions()
        } catch {
            // Use defaults if API fails
            suggestions = [
                "What did I spend this month?",
                "Show my top expenses",
                "What's my net worth?",
                "Create an invoice for..."
            ]
        }
    }

    // MARK: - Voice Actions

    /// Toggle the microphone on/off for speech-to-text.
    ///
    /// Behavior intentionally separates LISTEN from SEND so the user can
    /// review and edit before sending:
    ///   - Tap to START listening → clears the input field, starts capture.
    ///     The AIChatView mirrors `speechManager.transcribedText` into
    ///     `input` live (.onChange), so the TextField shows words as the
    ///     user speaks and the Send button activates naturally.
    ///   - Tap to STOP listening → leaves the final transcription in
    ///     `input`. Does NOT auto-send. User taps the Send arrow themselves.
    @MainActor
    func toggleMicrophone() async {
        if speechManager.isListening {
            speechManager.stopListening()
            // Make sure the final transcription is in `input` (the live mirror
            // .onChange in the view already does this, but be defensive).
            let final = speechManager.transcribedText.trimmingCharacters(in: .whitespacesAndNewlines)
            if !final.isEmpty {
                input = final
            }
            speechManager.resetTranscription()
        } else {
            // Request permissions if needed
            if !speechManager.hasMicrophonePermission || !speechManager.hasSpeechPermission {
                await speechManager.requestPermissions()
            }
            // Clear the field so the live mirror starts from a clean slate.
            input = ""
            speechManager.resetTranscription()
            speechManager.startListening()
        }
    }

    /// Toggle voice response (text-to-speech) on/off.
    @MainActor
    func toggleVoiceResponse() {
        isVoiceResponseEnabled.toggle()
        if !isVoiceResponseEnabled {
            speechManager.stopSpeaking()
        }
    }

    /// Enter or exit fullscreen voice mode.
    @MainActor
    func toggleVoiceMode() async {
        if isVoiceModeActive {
            exitVoiceMode()
        } else {
            await enterVoiceMode()
        }
    }

    /// Enter voice mode: enable TTS, start listening.
    @MainActor
    func enterVoiceMode() async {
        // Request permissions if needed
        if !speechManager.hasMicrophonePermission || !speechManager.hasSpeechPermission {
            await speechManager.requestPermissions()
        }

        guard speechManager.hasMicrophonePermission && speechManager.hasSpeechPermission else {
            errorMessage = "Microphone and speech recognition permissions are required for voice mode."
            return
        }

        isVoiceModeActive = true
        isVoiceResponseEnabled = true
        speechManager.resetTranscription()
        speechManager.startListening()
    }

    /// Exit voice mode: stop listening, stop speaking.
    @MainActor
    func exitVoiceMode() {
        isVoiceModeActive = false
        speechManager.stopListening()
        speechManager.stopSpeaking()
        speechManager.resetTranscription()
    }

    /// In voice mode, send the current transcription and re-listen.
    @MainActor
    func sendVoiceInput() async {
        speechManager.stopListening()
        let text = speechManager.transcribedText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else {
            // Nothing transcribed; restart listening
            speechManager.resetTranscription()
            speechManager.startListening()
            return
        }
        input = text
        speechManager.resetTranscription()
        await sendMessage()
    }

    // MARK: - Helpers

    func dismissError() {
        errorMessage = nil
    }

    private func scrollToBottom() {
        scrollToBottomTrigger += 1
    }

    private func scrollToLastMessageTop() {
        scrollToLastMessageTopTrigger += 1
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/aichat/AIChatViewModel.kt
package com.spentyai.app.features.aichat

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.spentyai.app.core.models.ChatMessage
import com.spentyai.app.core.models.ChatRole
import com.spentyai.app.core.network.ApiResult
import com.spentyai.app.core.services.AndroidSpeechManager
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.UUID

data class AIChatUiState(
    val messages: List<ChatMessage> = emptyList(),
    val input: String = "",
    val isSending: Boolean = false,
    val suggestions: List<String> = emptyList(),
    val errorMessage: String? = null,
    val scrollToBottomTrigger: Int = 0,
    val showClearConfirmation: Boolean = false
)

class AIChatViewModel(
    private val repository: AIChatRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(AIChatUiState())
    val uiState: StateFlow<AIChatUiState> = _uiState.asStateFlow()

    val canSend: Boolean
        get() = _uiState.value.input.isNotBlank() && !_uiState.value.isSending

    // ─── Voice ────────────────────────────────────────────────────────────────

    val speechManager = AndroidSpeechManager()

    private val _isVoiceResponseEnabled = MutableStateFlow(false)
    val isVoiceResponseEnabled: StateFlow<Boolean> = _isVoiceResponseEnabled.asStateFlow()

    private val _isVoiceModeActive = MutableStateFlow(false)
    val isVoiceModeActive: StateFlow<Boolean> = _isVoiceModeActive.asStateFlow()

    // ─── Chat ─────────────────────────────────────────────────────────────────

    fun loadSuggestions() {
        _uiState.update { it.copy(suggestions = repository.getDefaultSuggestions()) }
    }

    fun loadHistory() {
        viewModelScope.launch {
            when (val result = repository.loadHistory()) {
                is ApiResult.Success -> {
                    _uiState.update {
                        it.copy(
                            messages = result.data,
                            scrollToBottomTrigger = it.scrollToBottomTrigger + 1
                        )
                    }
                }
                is ApiResult.Failure -> {
                    // Loading history is OPPORTUNISTIC. If it fails (network blip,
                    // first-time user with no history yet, transient hiccup, etc.)
                    // the user can still send messages — surfacing a blocking
                    // alert here was over-aggressive and hid the chat UI.
                    // Log for debugging; leave messages empty.
                    android.util.Log.w("AIChat", "loadHistory failed: ${result.error}")
                }
            }
        }
    }

    fun onInputChange(text: String) {
        _uiState.update { it.copy(input = text) }
    }

    fun sendMessage(context: Context? = null) {
        // Reentry guard: a fast double-tap on Send or a voice-mode tap
        // arriving while a previous request is still in flight would queue
        // two requests and double-fire. Protect at the source — UI redraw
        // of the .disabled state has a frame of latency.
        if (_uiState.value.isSending) {
            android.util.Log.w("AIChat", "sendMessage ignored — already sending")
            return
        }

        // Belt-and-suspenders: if `input` is empty but the user is mid-
        // dictation (mic on with a transcript), fall through to the
        // transcribed text and also stop the mic. Catches a race where the
        // user taps Send before the live-mirror LaunchedEffect has
        // propagated the latest transcribedText into uiState.input.
        if (_uiState.value.input.trim().isEmpty()) {
            val fallback = speechManager.transcribedText.value.trim()
            if (fallback.isNotEmpty()) {
                android.util.Log.i("AIChat", "sendMessage rescuing from transcribedText (input was empty)")
                if (speechManager.isListening.value) {
                    speechManager.stopListening()
                }
                _uiState.update { it.copy(input = fallback) }
                speechManager.resetTranscription()
            }
        }

        val text = _uiState.value.input.trim()
        if (text.isEmpty()) {
            android.util.Log.w("AIChat", "sendMessage ignored — input is empty after fallback")
            return
        }

        android.util.Log.d("AIChat", "sendMessage start — ${text.take(60)}")

        val userMessage = ChatMessage(
            id = UUID.randomUUID().toString(),
            role = ChatRole.USER,
            content = text
        )

        _uiState.update {
            it.copy(
                messages = it.messages + userMessage,
                input = "",
                isSending = true,
                errorMessage = null,
                scrollToBottomTrigger = it.scrollToBottomTrigger + 1
            )
        }

        viewModelScope.launch {
            when (val result = repository.sendMessage(text, _uiState.value.messages, voiceMode = _isVoiceModeActive.value)) {
                is ApiResult.Success -> {
                    _uiState.update {
                        it.copy(
                            messages = it.messages + result.data,
                            isSending = false,
                            scrollToBottomTrigger = it.scrollToBottomTrigger + 1
                        )
                    }
                    // If the AI just created a financial account, push a
                    // refresh event so any open AccountListScreen reloads
                    // without the user having to pull-to-refresh. Mirrors
                    // iOS's .accountsDidChange NotificationCenter post.
                    if (result.data.accountCreated == true) {
                        com.spentyai.app.core.events.AppEvents.notifyAccountsChanged()
                    }
                    // Speak the AI response if voice response or voice mode is active.
                    if (context != null && (_isVoiceResponseEnabled.value || _isVoiceModeActive.value)) {
                        val content = result.data.content
                        // In voice mode, after speaking restart listening.
                        if (_isVoiceModeActive.value) {
                            speechManager.onTtsFinished = {
                                if (_isVoiceModeActive.value) {
                                    speechManager.startListening(context)
                                }
                            }
                        }
                        speechManager.speak(content, context)
                    }
                }
                is ApiResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isSending = false,
                            errorMessage = "Failed to get a response. Please try again."
                        )
                    }
                }
            }
        }
    }

    fun sendSuggestion(suggestion: String) {
        _uiState.update { it.copy(input = suggestion) }
        sendMessage()
    }

    fun clearHistory() {
        viewModelScope.launch {
            repository.clearHistory()
            _uiState.update {
                it.copy(
                    messages = emptyList(),
                    errorMessage = null,
                    showClearConfirmation = false
                )
            }
        }
    }

    fun showClearConfirmation() {
        _uiState.update { it.copy(showClearConfirmation = true) }
    }

    fun dismissClearConfirmation() {
        _uiState.update { it.copy(showClearConfirmation = false) }
    }

    fun dismissError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    // ─── Voice actions ────────────────────────────────────────────────────────

    /**
     * Toggle the microphone on/off.
     * If currently listening: stop and send the transcribed text if any.
     * If not listening: start listening (caller must have already obtained the permission).
     */
    suspend fun toggleMicrophone(context: Context) {
        if (speechManager.isListening.value) {
            speechManager.stopListening()
            // Finalize transcription into the input field but do NOT auto-send.
            // User reviews and taps Send themselves. Live-mirror in the screen
            // also keeps input in sync while listening.
            val transcribed = speechManager.transcribedText.value.trim()
            if (transcribed.isNotEmpty()) {
                _uiState.update { it.copy(input = transcribed) }
            }
            speechManager.resetTranscription()
        } else {
            speechManager.checkPermissions(context)
            if (speechManager.hasMicrophonePermission.value) {
                // Clear the field so live-mirror starts from a clean slate.
                _uiState.update { it.copy(input = "") }
                speechManager.resetTranscription()
                speechManager.startListening(context)
            }
        }
    }

    /** Toggle TTS for AI responses (voice response mode). */
    fun toggleVoiceResponse() {
        _isVoiceResponseEnabled.value = !_isVoiceResponseEnabled.value
    }

    /** Enter or exit voice mode. */
    suspend fun toggleVoiceMode(context: Context) {
        if (_isVoiceModeActive.value) {
            exitVoiceMode()
        } else {
            enterVoiceMode(context)
        }
    }

    /** Enter voice mode: check permissions, activate, start listening. */
    suspend fun enterVoiceMode(context: Context) {
        speechManager.checkPermissions(context)
        if (!speechManager.hasMicrophonePermission.value) {
            // Caller (Composable) is responsible for requesting the runtime permission.
            // We set a flag so the UI knows to show the permission rationale.
            return
        }
        _isVoiceModeActive.value = true
        speechManager.resetTranscription()
        speechManager.startListening(context)
    }

    /** Exit voice mode: stop all audio activity. */
    fun exitVoiceMode() {
        _isVoiceModeActive.value = false
        speechManager.stopListening()
        speechManager.stopSpeaking()
        speechManager.onTtsFinished = null
        speechManager.resetTranscription()
    }

    /**
     * Send the current voice input, or restart listening if the transcript is empty.
     * Used by the "Send" button inside the voice mode overlay.
     */
    suspend fun sendVoiceInput(context: Context) {
        speechManager.stopListening()
        val transcribed = speechManager.transcribedText.value.trim()
        if (transcribed.isNotEmpty()) {
            _uiState.update { it.copy(input = transcribed) }
            speechManager.resetTranscription()
            sendMessage(context)
        } else {
            // Nothing to send — restart listening.
            speechManager.startListening(context)
        }
    }

    // ─── Lifecycle ────────────────────────────────────────────────────────────

    override fun onCleared() {
        super.onCleared()
        speechManager.destroy()
    }
}
